java-client/session/src/main/java/io/deephaven/client/impl/BatchTableRequestBuilder.java

```diff
@@ -0,0 +1,582 @@
+                    // we've already exported it, no need to include it in our batch request
+                    continue;
+                }
+                // it's a new export
+            } else {
+                if (newExports.contains(tableSpec)) {
+                    throw new IllegalStateException(
+                        "Found a \"new export\", but it was not present in ExportLookup");
+                }
+                // is not an export, so we need to provide an empty ticket
+                ticket = Ticket.getDefaultInstance();
+            }
+
+            final Operation operation =
+                tableSpec.walk(new OperationAdapter(ticket, indices, lookup)).getOut();
+            builder.addOps(operation);
+            indices.put(tableSpec, ix++);
+        }
+
+        return builder.build();
+    }
+
+    private static <T> Operation op(BiFunction<Builder, T, Builder> f, T value) {
+        return f.apply(Operation.newBuilder(), value).build();
+    }
+
+    private static class OperationAdapter implements TableSpec.Visitor {
+        private final Ticket ticket;
+        private final Map<TableSpec, Integer> indices;
+        private final ExportLookup lookup;
+        private Operation out;
+
+        OperationAdapter(Ticket ticket, Map<TableSpec, Integer> indices, ExportLookup lookup) {
+            this.ticket = Objects.requireNonNull(ticket);
+            this.indices = Objects.requireNonNull(indices);
+            this.lookup = Objects.requireNonNull(lookup);
+        }
+
+        public Operation getOut() {
+            return Objects.requireNonNull(out);
+        }
+
+        private TableReference ref(TableSpec table) {
+            Optional<Ticket> existing = lookup.ticket(table);
+            if (existing.isPresent()) {
+                return TableReference.newBuilder().setTicket(existing.get()).build();
+            }
+            final Integer ix = indices.get(table);
+            if (ix != null) {
+                return TableReference.newBuilder().setBatchOffset(ix).build();
+            }
+            throw new IllegalStateException(
+                "Unable to reference table - batch table request logic has a bug.");
+        }
+
+        @Override
+        public void visit(EmptyTable emptyTable) {
+            out = op(Builder::setEmptyTable,
+                EmptyTableRequest.newBuilder().setResultId(ticket).setSize(emptyTable.size()));
+        }
+
+        @Override
+        public void visit(NewTable newTable) {
+            throw new UnsupportedOperationException("TODO");
+        }
+
+        @Override
+        public void visit(TimeTable timeTable) {
+            // noinspection Convert2Lambda
+            timeTable.timeProvider().walk(new Visitor() {
+                @Override
+                public void visit(TimeProviderSystem system) {
+                    // OK
+                    // note: if other time providers are added, we need to throw an exception
```

Comment:
this comment could use improvement; I cannot tell if it is a TODO or a "by the way; in the event ..." etc.

Reply:
Hopefully clarified.